torchrl/envs/custom/llm.py

```diff
@@ -4,23 +4,252 @@
+        # self.action_key = unravel_key(action_key)
+        if str2str:
+            self.observation_spec = Composite(
+                {
+                    observation_key: NonTensor(
+                        example_data="a string", batched=True, shape=()
+                    )
+                }
+            )
+            self.action_spec = Composite(
+                {action_key: NonTensor(example_data="a string", batched=True, shape=())}
+            )
+        else:
+            if vocab_size is None:
+                self.observation_spec = Composite(
+                    {
+                        observation_key: Unbounded(
+                            shape=(-1,), dtype=torch.int64, device=device
+                        )
+                    }
+                )
+                self.action_spec = Composite(
+                    {
+                        action_key: Unbounded(
+                            shape=(-1,), dtype=torch.int64, device=device
+                        )
+                    }
+                )
+            else:
+                self.observation_spec = Composite(
+                    {
+                        observation_key: Bounded(
+                            shape=(-1,),
+                            dtype=torch.int64,
+                            low=0,
+                            high=vocab_size,
+                            device=device,
+                        )
+                    }
+                )
+                self.action_spec = Composite(
+                    {
+                        action_key: Bounded(
+                            shape=(-1,),
+                            dtype=torch.int64,
+                            low=0,
+                            high=vocab_size,
+                            device=device,
+                        )
+                    }
+                )
+        self.full_done_spec = Composite(
+            done=Unbounded(shape=(1,), dtype=torch.bool),
+            truncated=Unbounded(shape=(1,), dtype=torch.bool),
+            terminated=Unbounded(shape=(1,), dtype=torch.bool),
+        )
+
+    @classmethod
+    def from_dataloader(
+        cls,
+        dataloader: DataLoader,
+        *,
+        observation_key: NestedKey = "observation",
+        action_key: NestedKey = "action",
+        str2str: bool = False,
+        device: torch.device | None = None,
+        vocab_size: int | None = None,
+        primers: Composite | None = None,
+        data_keys: list[NestedKey] | None = None,
+        data_specs: list[TensorSpec] | None = None,
+        example_data: Any = None,
+        stack_method: Callable[[Any], Any]
+        | Literal["as_nested_tensor", "as_padded_tensor"] = None,
+    ) -> LLMEnv:
+        """Creates an LLMEnv instance from a dataloader.
+
+        This method creates an LLMEnv instance and appends a DataLoadingPrimer to it, which loads data from the provided dataloader.
```

Comment:
From a n00b perspective, this would help me when reading this

```suggestion
        This method creates an LLMEnv instance and appends a DataLoadingPrimer to it, which populates ``data_keys`` (by default ``observation_key``) with data from the provided dataloader when the environment is reset.
```